state-chain/pallets/cf-broadcast/src/migrations/v3.rs

```diff
@@ -0,0 +1,283 @@
+use crate::*;
+#[cfg(feature = "try-runtime")]
+use frame_support::sp_runtime::DispatchError;
+use frame_support::{traits::OnRuntimeUpgrade, weights::Weight};
+#[cfg(feature = "try-runtime")]
+use sp_std::prelude::Vec;
+use sp_std::{collections::btree_map::BTreeMap, marker::PhantomData};
+
+pub(crate) mod old {
+	use super::*;
+
+	/// A unique id for each broadcast attempt
+	#[derive(
+		Clone, Debug, PartialEq, Eq, Encode, Decode, TypeInfo, MaxEncodedLen, Default, Copy,
+	)]
+	pub struct BroadcastAttemptId {
+		pub broadcast_id: BroadcastId,
+		pub attempt_count: AttemptCount,
+	}
+
+	#[derive(Clone, RuntimeDebug, PartialEq, Eq, Encode, Decode, TypeInfo)]
+	pub struct BroadcastAttempt<T: Config<I>, I: 'static> {
+		pub broadcast_attempt_id: BroadcastAttemptId,
+		pub transaction_payload: TransactionFor<T, I>,
+		pub threshold_signature_payload: PayloadFor<T, I>,
+		pub transaction_out_id: TransactionOutIdFor<T, I>,
+	}
+
+	#[derive(Clone, RuntimeDebug, PartialEq, Eq, Encode, Decode, TypeInfo)]
+	pub struct TransactionSigningAttempt<T: Config<I>, I: 'static> {
+		pub broadcast_attempt: BroadcastAttempt<T, I>,
+		pub nominee: T::ValidatorId,
+	}
+
+	#[frame_support::storage_alias]
+	pub type AwaitingBroadcast<T: Config<I>, I: 'static> =
+		StorageMap<Pallet<T, I>, Twox64Concat, BroadcastAttemptId, TransactionSigningAttempt<T, I>>;
+
+	#[frame_support::storage_alias]
+	pub type FailedBroadcasters<T: Config<I>, I: 'static> =
+		StorageMap<Pallet<T, I>, Twox64Concat, BroadcastId, Vec<<T as Chainflip>::ValidatorId>>;
+
+	#[frame_support::storage_alias]
+	pub type BroadcastRetryQueue<T: Config<I>, I: 'static> =
+		StorageValue<Pallet<T, I>, Vec<BroadcastAttempt<T, I>>>;
+}
+
+#[derive(Clone, Encode, Decode)]
+pub struct MigrationVerification<T: Chainflip> {
+	broadcasts_with_failed_broadcasters:
+		Vec<(BroadcastId, BTreeSet<<T as Chainflip>::ValidatorId>)>,
+	broadcast_retry_queue: Vec<BroadcastId>,
+}
+pub struct Migration<T: Config<I>, I: 'static>(PhantomData<(T, I)>);
+
+impl<T: Config<I>, I: 'static> OnRuntimeUpgrade for Migration<T, I> {
+	fn on_runtime_upgrade() -> frame_support::weights::Weight {
+		// Awaiting broadcast: take the broadcast attempt with the highest attempt.
+		let mut latest_attempt = BTreeMap::new();
+		old::AwaitingBroadcast::<T, I>::iter_keys().for_each(
+			|old::BroadcastAttemptId { broadcast_id, attempt_count }| {
+				if match latest_attempt.get(&broadcast_id) {
+					Some(attempt) => attempt_count > *attempt,
+					None => true,
+				} {
+					latest_attempt.insert(broadcast_id, attempt_count);
+				}
+			},
+		);
+
+		// Migrate data from old storage into new storage.
+		latest_attempt.into_iter().for_each(|(broadcast_id, attempt_count)| {
+			if let Some(attempt) = old::AwaitingBroadcast::<T, I>::get(old::BroadcastAttemptId {
```

Comment:
The old AwaitingBroadcast storage needs to be removed. You could use `drain` to store all the old::SigningAttempts into a BTreeMap above and then write them to the new storage here.